cpp/src/umap/simpl_set_embed/algo.h

```diff
@@ -136,12 +135,13 @@ __global__ void optimize_batch_kernel(
   const int *tail, int nnz, T *epochs_per_sample, int n_vertices,
   bool move_other, T *epochs_per_negative_sample,
   T *epoch_of_next_negative_sample, T *epoch_of_next_sample, double alpha,
-  int epoch, double gamma, uint64_t seed, UMAPParams params) {
+  int epoch, double gamma, uint64_t seed, double* embedding_updates, UMAPParams params) {
```

Comment:
I believe the associativity problem with the non-determinism of floating-point updates is actually a bug and we should keep that code to always execute. If the rounding and the use of double fixes this issue, perhaps we should have the algorithm always return double outputs. What do you think?

Reply:
Absolutely, the multi-core execution implementation probably also suffers from the bug, even if it is less visible, and will benefit from storing the embeddings themselves in doubles.